### DISCUSSION TITLE
Angular Compatibility

### DISCUSSION THREAD

**Discussion by kyaneva** [umami-software/umami#1572] [Q&A]
Hello,

is there a plugin for Angular projects, so that the Event tracking using JS would work?

If not, is it somehow possible, to track the page views other than '/', because currently only the home page is beeing displayed and all views are considered to be on the home page?

**Comment by briancao** [comment]
You can possibly disable the auto-track during import ( `<script ... data-auto-track="false"></script>`) and fire off the tracker manually in angular using the tracker functions in JS.

https://umami.is/docs/tracker-functions

I believe Angular has router middleware you can tap into to fire off the tracker function.

**Comment by kyaneva** [comment]
The problem is that there is no umami instance on which I could call a tracker function. That is because there is no npm  package for umami that is compatible with Angular. Or am I wrong?

The page views are actually in the db in the pageviews table, but they are not shown on the dashboard under Pages.

<img width="560" alt="Screenshot 2022-10-20 at 09 45 02" src="https://user-images.githubusercontent.com/33027135/196887752-73fcc042-8d81-478b-a3bf-81841602ddf9.png">

Only under Realtime logs:
<img width="311" alt="Screenshot 2022-10-20 at 09 48 13" src="https://user-images.githubusercontent.com/33027135/196888417-3bdd6a66-20cd-45ae-9d29-a57932b9e397.png">

**Reply by briancao** [reply]
Correct, there is no specific package or solution that we have for Angular.  You'll have to come up with a custom solution, like the one I  recommended.  


Or maybe have a top level subscription to the router and custom send events from that.  Something to the nature of this pseudo code.
```
import * as umami from './assets/umami'; // or however you're importing umami

export class AppComponent {

    constructor(private router: Router) {
        //Router subscriber
        this.router.events.subscribe((event: Event) => {
            if (event instanceof NavigationStart) {
                umami.trackView(event.url, 'https://www.google.com');
            }
       }
            
```

I haven't touched angular in years, so please treat this as a high level example.

**Reply by kyaneva** [reply]
The problem itself was, that I could not import the umami instance in any kind of way, so I came up with the solution to use window.umami as an umami instance. Thank you for the suggestions!

**Reply by konstantinosftw** [reply]
> The problem itself was, that I could not import the umami instance in any kind of way, so I came up with the solution to use window.umami as an umami instance. Thank you for the suggestions!

Hello @kyaneva! How did you manage to import the umami instance in Angular? When I write `window.umami` in my .ts file I get `Property 'umami' does not exist on type 'window'`.

**Reply by FHU-yezi** [reply]
> > The problem itself was, that I could not import the umami instance in any kind of way, so I came up with the solution to use window.umami as an umami instance. Thank you for the suggestions!
> 
> Hello @kyaneva! How did you manage to import the umami instance in Angular? When I write `window.umami` in my .ts file I get `Property 'umami' does not exist on type 'window'`.

Just ignore it.

**Reply by konstantinosftw** [reply]
> Just ignore it.

Thanks @FHU-yezi for the answer. I tried ignoring it, but the app does not compile because of it. It feels like I should be importing something.

<img src="https://github.com/umami-software/umami/assets/23435302/0ce9f029-7590-4741-bb19-5697c19832e8"  width="80%">

**Reply by FHU-yezi** [reply]
Maybe use `umami.track()` ( remove `window.` ) and add `// @ts-ignore` before the track code line? They should make the compile success.

**Reply by konstantinosftw** [reply]
@FHU-yezi thanks a lot, you nailed it. This worked 🔥

```
// @ts-ignore
umami.track()
```
